proposals/p3927.md

```diff
@@ -0,0 +1,132 @@
+# More consistent package syntax
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/3927)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Status quo](#status-quo)
+
+<!-- tocstop -->
+
+## Abstract
+
+Change the syntax for `package` declarations from:
+
+```carbon
+[package Foo] [library "bar"] api;
+[package Foo] [library "bar"] impl;
+```
+
+to
+
+```carbon
+[package Foo] [library "bar"];
+impl [package Foo] [library "bar"];
+```
+
+## Problem
+
+The `package` declaration is currently inconsistent with other Carbon
+declarations:
+
+-   Modifier keywords for other declarations precede the introducer keyword.
+    However, for package declarations, the `api` / `impl` modifier appears at
+    the end of the declaration.
```

Comment:
FWIW, this part I have mixed feelings about. It had seemed really nice for us to have a reliably introductory string at the top of the file, and we somewhat lose that.

Maybe we already lost that with the default package?

That said, it seems like this should really go before `library`, not `package` here?

Reply:
> It had seemed really nice for us to have a reliably introductory string at the top of the file

Is that from an aesthetic perspective, or are there technical advantages to having a fixed initial token sequence that's shared across all files in a library that we should be considering here?

FWIW, having made the change across our documentation for this PR, I find that I do like how `impl package Foo library "bar";` reads -- the first thing that tells me is the kind of file that I'm in, which seems pretty useful, whereas putting it at the end feels more like an afterthought or ignorable detail.

> That said, it seems like this should really go before `library`, not `package` here?

Like so?
```carbon
package Foo impl library "bar";
package Foo impl;
impl library "bar";
```
I can see the logic, but it feels less satisfying to me than the status quo or the proposed approach.

Reply:
It might also be worth considering how a prefix looks with imports:

```
impl package Foo library "wiz";

import package Pitter library "patter";
import library "bar";
export import library "baz";
```

The positioning is more consistent as a prefix, and not requiring `api` feels like an incremental writability improvement. I expect we would not insert a keyword before `library` inside an `import`, but we do put `import` before `package`.

Reply:
I was a bit worried about tools that want to look for a magic initial sequence, but I don't think that's too important. This was mostly an aesthetic question for me.

The comparison with `import` is a really nice point though, and seems to really clearly show my idea about `library` was wrong. Especially given that this *does* do an implicit import of exactly the API named with the string following `impl`.

Ok, I think I'm sold on the leading `impl`. =]

Reply:
Added an "Alternative considered" for putting `impl` immediately before where `library` is / would be.

Updated the "Alternatives considered" to mention @jonmeow's observation about consistency with `import`.